src/adaptors/tcp/tcp_adaptor.c

```diff
@@ -477,43 +722,42 @@ static int handle_incoming(qdr_tcp_connection_t *conn, const char *msg)
         assert (DEQ_SIZE(buffers) == 0);
     }
 
-    // Close the stream message if read side has closed
-    if (IS_ATOMIC_FLAG_SET(&conn->raw_closed_read)) {
-        qd_log(log, QD_LOG_DEBUG,
-            DLV_FMT" close %s in_dlv_stream delivery",
-            DLV_ARGS(conn->in_dlv_stream), qdr_tcp_connection_role_name(conn));
-        qd_message_set_receive_complete(qdr_delivery_message(conn->in_dlv_stream));
-        qdr_delivery_continue(tcp_adaptor->core, conn->in_dlv_stream, true);
-        conn->raw_read_shutdown = true;
-    }
+    check_handle_incoming_closed_read(conn);
 
     return count;
 }
 
-static void free_qdr_tcp_connection(qdr_tcp_connection_t *tc)
+static void free_qdr_tcp_connection(qdr_tcp_connection_t *conn)
 {
-    qdr_tcp_stats_t *tcp_stats = get_tcp_stats(tc);
+    qdr_tcp_stats_t *tcp_stats = get_tcp_stats(conn);
     LOCK(&tcp_stats->stats_lock);
     tcp_stats->connections_closed += 1;
     UNLOCK(&tcp_stats->stats_lock);
 
-    qd_tcp_connector_decref(tc->connector);
-    qd_tcp_listener_decref(tc->listener);
-    vflow_end_record(tc->vflow);
-    free(tc->reply_to);
-    free(tc->remote_address);
-    free(tc->global_id);
-    sys_atomic_destroy(&tc->q2_restart);
-    sys_atomic_destroy(&tc->raw_closed_read);
-    sys_atomic_destroy(&tc->raw_closed_write);
-    qd_timer_free(tc->activate_timer);
-    sys_mutex_free(&tc->activation_lock);
-    free(tc->write_buffer.bytes);
-    free(tc->read_buffer.bytes);
-    //proactor will free the socket
+    qd_tcp_connector_decref(conn->connector);
+    qd_tcp_listener_decref(conn->listener);
+    vflow_end_record(conn->vflow);
+    free(conn->reply_to);
+    free(conn->remote_address);
+    free(conn->global_id);
+    sys_atomic_destroy(&conn->q2_restart);
+    sys_atomic_destroy(&conn->raw_closed_read);
+    sys_atomic_destroy(&conn->raw_closed_write);
+    qd_timer_free(conn->activate_timer);
+    sys_mutex_free(&conn->activation_lock);
+    free_read_write_adaptor_buffers(conn);
 
-    qd_free_tcp_adaptor_config(tc->config, tcp_adaptor->log_source);
-    free_qdr_tcp_connection_t(tc);
+    // Free tls related stuff if need be.
+    if (conn->tls) {
+        if (conn->tls->tls_session)
+            qd_tls_stop(conn->tls);
+        free_qd_tls_t(conn->tls);
```

Comment:
If conn->tls cannot be re-used (restarted?) after calling qd_tls_stop() then qd_tls_stop() should free conn->tls, not the adaptor.